File: python/docsearch/nav.py
"""How a documentation site's pages nest.

Hierarchy only. Which pages exist is a separate question with separate sources,
answered in :mod:`docsearch.discover`. Keeping them apart is what the probed
targets forced: a rendered sidebar is not a page set. One target has none at
all, and the other renders 19 links against 210 pages because its generator
collapses categories client-side.

So a hierarchy source is *checked against coverage before it is believed*. A
sidebar that places a tenth of the site is not the site's structure, whatever
it looks like, and the pages no source mentions are placed by their URL path
and reported rather than dropped -- dropping them discards 41 of one target's
210 pages, including every command reference.

The output is what the chunker needs: for each page, a dotted section number
and the heading ancestry above it.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from urllib.parse import urlsplit

from bs4 import BeautifulSoup
from bs4.element import Tag

from .fetch import normalize

__all__ = ["Hierarchy", "Placement", "derive", "index_page_tree", "sidebar_tree"]

#: Share of the page set a sidebar must place before it is treated as the
#: site's structure. Below this it is a fragment of a navigation, not a map of
#: one -- the case a collapsed generator sidebar produces.
MIN_COVERAGE = 0.5

#: Elements that plausibly hold a documentation navigation.
_NAV_SELECTOR = (
    "nav, aside, [class*=sidebar], [class*=Sidebar], [class*=toc], [class*=menu], [id*=sidebar]"
)

_HEADINGS = ("h1", "h2", "h3", "h4", "h5", "h6")


@dataclass(frozen=True, slots=True)
class Placement:
    """Where one page sits in the tree."""

    url: str
    #: Dotted position, e.g. "3.2". Becomes the chunk's authoritative section.
    section: str
    #: Heading ancestry above the page, root first.
    ancestry: tuple[str, ...]
    title: str


@dataclass(slots=True)
class Hierarchy:
    placements: list[Placement] = field(default_factory=list)
    #: 'sidebar_dom' | 'index_page' | 'url_path'
    source: str = "url_path"
    #: True when the structure was inferred rather than declared, so nothing
    #: exists to check it against.
    inferred: bool = True
    #: Pages no declared source mentioned, placed by URL path instead.
    placed_by_path: list[str] = field(default_factory=list)
    notes: list[str] = field(default_factory=list)

    def by_url(self) -> dict[str, Placement]:
        return {p.url: p for p in self.placements}


@dataclass(slots=True)
class _Node:
    title: str
    url: str | None = None
    children: list[_Node] = field(default_factory=list)


def _host(url: str) -> str:
    return (urlsplit(url).hostname or "").lower()


def _abs(base: str, href: str) -> str:
    from urllib.parse import urljoin

    return normalize(urljoin(base, href))


def _usable_href(tag: Tag) -> str | None:
    href = str(tag.get("href", "")).strip()
    if not href or href.startswith(("#", "mailto:", "javascript:", "tel:")):
        return None
    return href


# -- source: a rendered sidebar -------------------------------------------


def sidebar_tree(html: bytes, base: str) -> list[_Node]:
    """The richest navigation the page renders, as a tree.

    Chooses the candidate container holding the most same-host links rather
    than the first that matches, because a marketing header is a ``<nav>`` too.
    Whether the result is worth believing is decided by coverage, not here.
    """
    soup = BeautifulSoup(html, "lxml")
    for tag in soup(["script", "style"]):
        tag.decompose()

    best: Tag | None = None
    best_links = 0
    for cand in (soup.body or soup).select(_NAV_SELECTOR):
        links = [
            a
            for a in cand.find_all("a", href=True)
            if _usable_href(a) and _host(_abs(base, str(a["href"]))) == _host(base)
        ]
        if len(links) > best_links:
            best, best_links = cand, len(links)
    if best is None or best_links == 0:
        return []
    return _tree_from_lists(best, base)


def _own_anchor(li: Tag) -> Tag | None:
    """The link belonging to this item, not to one of its children.

    ``find`` searches descendants, so a category item wrapping a sublist would
    otherwise adopt its first child's link -- taking that child's title and URL
    for itself and then listing the child again beneath it.
    """
    for a in li.find_all("a", href=True):
        holder = a.find_parent(["ul", "ol"])
        if holder is not None and holder is not li.parent:
            continue
        return a if isinstance(a, Tag) else None
    return None


def _tree_from_lists(container: Tag, base: str) -> list[_Node]:
    """Build a tree from nested ``<ul>``/``<li>``, or a flat list if unnested."""

    def walk(scope: Tag, *, top: bool = False) -> list[_Node]:
        out: list[_Node] = []
        items: list[Tag] = []
        for lst in scope.find_all(["ul", "ol"], recursive=False):
            items.extend(lst.find_all("li", recursive=False))
        if not items:
            if not top:
                # A leaf item has no children. Falling through to the flat
                # branch would find the item's own link again and hang a copy
                # of the page beneath itself.
                return out
            # A navigation with no list markup at all: take the links in
            # document order, flat.
            for a in scope.find_all("a", href=True):
                href = _usable_href(a)
                if href is None:
                    continue
                out.append(_Node(title=a.get_text(" ", strip=True), url=_abs(base, href)))
            return out

        for li in items:
            anchor = _own_anchor(li)
            href = _usable_href(anchor) if anchor is not None else None
            if anchor is not None and href is not None:
                title = anchor.get_text(" ", strip=True)
                node = _Node(title=title, url=_abs(base, href))
            else:
                # A category label with no link of its own still names a level.
                label = li.find(["span", "button", "div", *_HEADINGS])
                title = label.get_text(" ", strip=True) if label else li.get_text(" ", strip=True)
                node = _Node(title=title.split("\n")[0].strip())
            node.children = walk(li)
            out.append(node)
        return out

    return walk(container, top=True)


# -- source: a hub page's headings ----------------------------------------


def index_page_tree(html: bytes, base: str) -> list[_Node]:
    """Group a hub page's links under the headings above them.

    A heading *inside* a link titles that link; a heading outside one opens a
    section. Both shapes appear on the same real page -- cards are written as
    ``<a><h3>Title</h3></a>`` under an ``<h1>`` section header -- and treating
    every heading as a section opener assigns each link the title of the one
    before it, mislabelling the whole page in a way only a slug comparison
    reveals.
    """
    soup = BeautifulSoup(html, "lxml")
    for tag in soup(["script", "style", "nav", "footer"]):
        tag.decompose()

    roots: list[_Node] = []
    stack: list[tuple[int, _Node]] = []
    seen: set[str] = set()

    for el in (soup.body or soup).find_all([*_HEADINGS, "a"]):
        if el.name == "a":
            href = _usable_href(el)
            if href is None:
                continue
            url = _abs(base, href)
            if _host(url) != _host(base) or url in seen:
                continue
            seen.add(url)
            inner = el.find(_HEADINGS)
            title = (inner or el).get_text(" ", strip=True)
            node = _Node(title=title, url=url)
            (stack[-1][1].children if stack else roots).append(node)
            continue

        if el.find_parent("a") is not None:
            # Titles a link; already consumed above.
            continue
        level = int(el.name[1])
        while stack and stack[-1][0] >= level:
            stack.pop()
        node = _Node(title=el.get_text(" ", strip=True))
        (stack[-1][1].children if stack else roots).append(node)
        stack.append((level, node))

    return roots


# -- source: the URL path --------------------------------------------------


def url_path_tree(urls: list[str], seed: str) -> list[_Node]:
    """Nest pages by their path segments below the seed.

    Inferred, with nothing to check it against, which is why it warns. It is
    still far better than a flat list: it keeps a command reference together
    instead of scattering it.
    """
    base_depth = len([s for s in urlsplit(seed).path.split("/") if s])
    roots: list[_Node] = []
    groups: dict[tuple[str, ...], _Node] = {}

    for url in urls:
        segments = [s for s in urlsplit(url).path.split("/") if s][base_depth:]
        if not segments:
            roots.append(_Node(title=_title_from_slug(urlsplit(url).path), url=url))
            continue
        parent_children = roots
        prefix: tuple[str, ...] = ()
        for seg in segments[:-1]:
            prefix += (seg,)
            group = groups.get(prefix)
            if group is None:
                group = _Node(title=_title_from_slug(seg))
                groups[prefix] = group
                parent_children.append(group)
            parent_children = group.children
        parent_children.append(_Node(title=_title_from_slug(segments[-1]), url=url))
    return roots


def _title_from_slug(slug: str) -> str:
    text = slug.strip("/").split("/")[-1]
    for suffix in (".html", ".htm", ".md"):
        text = text.removesuffix(suffix)
    return text.replace("-", " ").replace("_", " ").strip().title() or "Untitled"


# -- selection and numbering ----------------------------------------------


def _flatten(nodes: list[_Node], prefix: str = "") -> list[Placement]:
    out: list[Placement] = []

    def walk(items: list[_Node], path: str, ancestry: tuple[str, ...]) -> None:
        for i, node in enumerate(items, start=1):
            section = f"{path}.{i}" if path else str(i)
            if node.url is not None:
                out.append(
                    Placement(
                        url=node.url,
                        section=section,
                        ancestry=ancestry,
                        title=node.title or _title_from_slug(node.url),
                    )
                )
            walk(node.children, section, ancestry + ((node.title,) if node.title else ()))

    walk(nodes, prefix, ())
    # A navigation may list one page in more than one place. The first
    # position wins: a page with two section numbers is two documents to
    # anything that filters by section.
    seen: set[str] = set()
    unique: list[Placement] = []
    for p in out:
        if p.url in seen:
            continue
        seen.add(p.url)
        unique.append(p)
    return unique


def _covered(placements: list[Placement], coverage: list[str]) -> float:
    if not coverage:
        return 0.0
    placed = {p.url for p in placements}
    return len(placed & set(coverage)) / len(coverage)


def derive(
    coverage: list[str],
    seed: str,
    *,
    seed_html: bytes | None = None,
) -> Hierarchy:
    """Choose a hierarchy source, place every page, and say what happened."""
    hier = Hierarchy()
    if not coverage:
        return hier

    candidates: list[tuple[str, list[Placement]]] = []
    if seed_html is not None:
        for name, nodes in (
            ("sidebar_dom", sidebar_tree(seed_html, seed)),
            ("index_page", index_page_tree(seed_html, seed)),
        ):
            if nodes:
                candidates.append((name, _flatten(nodes)))

    best_name, best_placements, best_share = "url_path", [], 0.0
    for name, placements in candidates:
        share = _covered(placements, coverage)
        hier.notes.append(f"{name} places {share:.0%} of the page set")
        if share > best_share:
            best_name, best_placements, best_share = name, placements, share

    if best_share < MIN_COVERAGE:
        # A fragment of a navigation is not a map of one. This is the case a
        # generator that collapses its categories produces, and believing it
        # would index a tenth of the site under a confident-looking tree.
        if best_share:
            hier.notes.append(
                f"{best_name} places only {best_share:.0%} of the page set; "
                f"falling back to URL path depth"
            )
        hier.source = "url_path"
        hier.inferred = True
        hier.placements = _flatten(url_path_tree(coverage, seed))
        hier.placed_by_path = list(coverage)
        return hier

    wanted = set(coverage)
    kept = [p for p in best_placements if p.url in wanted]
    missing = [u for u in coverage if u not in {p.url for p in kept}]

    if missing:
        # Placed, not dropped. Excluding them discards 41 of one probed
        # target's 210 pages, including every command reference -- the silent
        # partial index the completeness gate exists to prevent.
        extra = _flatten(url_path_tree(missing, seed), prefix=str(len(kept) + 1))
        kept.extend(extra)
        hier.placed_by_path = missing
        hier.notes.append(f"{len(missing)} page(s) absent from {best_name}, placed by URL path")

    hier.source = best_name
    hier.inferred = False
    hier.placements = kept
    return hier

File: internal/site/nav/nav.go
// Package nav says how a documentation site's pages nest.
//
// Hierarchy only. Which pages exist is a separate question with separate
// sources, answered in internal/site/discover. Keeping them apart is what
// the probed sites forced: a rendered sidebar is not a page set. One site has
// none at all, and another renders 19 links against 210 pages because its
// generator collapses categories in the browser.
//
// So a hierarchy source is checked against coverage before it is believed. A
// sidebar that places a tenth of the site is not the site's structure
// whatever it looks like, and pages no source mentions are placed by their
// URL path and reported rather than dropped: dropping them discarded 41 of
// one site's 210 pages, including every command reference.
//
// The output is what the chunker needs: for each page, a dotted section
// number and the heading ancestry above it.
//
// Ported from python/docsearch/nav.py.
package nav

import (
	"fmt"
	"net/url"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"

	"github.com/bamsammich/docsearch/internal/adapter/html"
	"github.com/bamsammich/docsearch/internal/domain"
	"github.com/bamsammich/docsearch/internal/site/fetch"
)

// minCoverage is the share of the page set a sidebar must place before it is
// treated as the site's structure. Below it, the sidebar is a fragment of a
// navigation rather than a map of one, which is what a generator that
// collapses its categories renders.
const minCoverage = 0.5

// navSelector matches the elements that plausibly hold a documentation
// navigation.
const navSelector = "nav, aside, [class*=sidebar], [class*=Sidebar], " +
	"[class*=toc], [class*=menu], [id*=sidebar]"

const headingSelector = "h1, h2, h3, h4, h5, h6"

// Placement is where one page sits in the tree.
type Placement struct {
	URL string
	// Section is the dotted position, "3.2", which becomes the chunk's
	// authoritative section.
	Section string
	Title   string
	// Ancestry is the heading ancestry above the page, root first.
	Ancestry []string
}

// Hierarchy is every page's placement, and how it was arrived at.
type Hierarchy struct {
	Placements []Placement
	// PlacedByPath are the pages no declared source mentioned.
	PlacedByPath []string
	Notes        []string
	// Source is domain.SourceSidebarDOM, domain.SourceIndexPage or
	// domain.SourceURLPath.
	Source domain.StructureSource
	// Inferred is true when the structure was read from URLs rather than
	// declared, so nothing exists to check it against.
	Inferred bool
}

// ByURL indexes the placements.
func (h *Hierarchy) ByURL() map[string]Placement {
	out := make(map[string]Placement, len(h.Placements))
	for _, p := range h.Placements {
		out[p.URL] = p
	}
	return out
}

// node is one entry of a navigation tree.
type node struct {
	title string
	// url is empty for a category that names a level without being a page.
	url      string
	children []*node
}

// Derive chooses a hierarchy source, places every page, and says what
// happened. seedHTML is the seed page, where the crawl read one.
func Derive(coverage []string, seed string, seedHTML []byte) (*Hierarchy, error) {
	h := &Hierarchy{Source: domain.SourceURLPath, Inferred: true}
	if len(coverage) == 0 {
		return h, nil
	}
	best, err := bestCandidate(h, seedHTML, seed, coverage)
	if err != nil {
		return nil, err
	}
	if best.share < minCoverage {
		// A fragment of a navigation is not a map of one. Believing it would
		// index a tenth of the site under a confident-looking tree.
		if best.share > 0 {
			h.Notes = append(h.Notes, fmt.Sprintf(
				"%s places only %s of the page set; falling back to URL path depth",
				best.name, percent(best.share)))
		}
		h.Placements = flatten(urlPathTree(coverage, seed), "")
		h.PlacedByPath = coverage
		return h, nil
	}
	h.place(best, coverage, seed)
	return h, nil
}

// candidate is one hierarchy source and how much of the page set it places.
type candidate struct {
	placements []Placement
	share      float64
	name       domain.StructureSource
}

// bestCandidate reads the declared sources the seed page offers and returns
// whichever places most of the page set.
func bestCandidate(
	h *Hierarchy,
	seedHTML []byte,
	seed string,
	coverage []string,
) (candidate, error) {
	best := candidate{name: domain.SourceURLPath}
	if seedHTML == nil {
		return best, nil
	}
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(string(seedHTML)))
	if err != nil {
		return best, fmt.Errorf("parse %s: %w", seed, err)
	}
	for _, source := range []struct {
		nodes []*node
		name  domain.StructureSource
	}{
		{nodes: sidebarTree(doc, seed), name: domain.SourceSidebarDOM},
		{nodes: indexPageTree(doc, seed), name: domain.SourceIndexPage},
	} {
		if len(source.nodes) == 0 {
			continue
		}
		placements := flatten(source.nodes, "")
		share := covered(placements, coverage)
		h.Notes = append(
			h.Notes,
			fmt.Sprintf("%s places %s of the page set", source.name, percent(share)),
		)
		if share > best.share {
			best = candidate{name: source.name, placements: placements, share: share}
		}
	}
	return best, nil
}

// place keeps the placements for pages the crawl knows, and places the rest
// by URL path. Excluding them would discard pages no navigation mentions,
// which is the silent partial index the completeness gate exists to prevent.
func (h *Hierarchy) place(best candidate, coverage []string, seed string) {
	wanted := make(map[string]bool, len(coverage))
	for _, u := range coverage {
		wanted[u] = true
	}
	var kept []Placement
	placed := map[string]bool{}
	for _, p := range best.placements {
		if wanted[p.URL] {
			kept = append(kept, p)
			placed[p.URL] = true
		}
	}
	var missing []string
	for _, u := range coverage {
		if !placed[u] {
			missing = append(missing, u)
		}
	}
	if len(missing) > 0 {
		kept = append(kept, flatten(urlPathTree(missing, seed), strconv.Itoa(len(kept)+1))...)
		h.PlacedByPath = missing
		h.Notes = append(h.Notes, fmt.Sprintf("%d page(s) absent from %s, placed by URL path",
			len(missing), best.name))
	}
	h.Source, h.Inferred, h.Placements = best.name, false, kept
}

// sidebarTree is the richest navigation the page renders, as a tree.
//
// The candidate container holding the most same-host links wins, rather than
// the first that matches, because a marketing header is a nav too. Whether
// the result is worth believing is decided by coverage, not here.
func sidebarTree(doc *goquery.Document, base string) []*node {
	doc = withoutScripts(doc)
	var best *goquery.Selection
	bestLinks := 0
	root(doc).Find(navSelector).Each(func(_ int, cand *goquery.Selection) {
		if n := sameHostLinks(cand, base); n > bestLinks {
			best, bestLinks = cand, n
		}
	})
	if best == nil {
		return nil
	}
	return treeFromLists(best, base, true)
}

// sameHostLinks counts the links in a container that address the same host.
func sameHostLinks(sel *goquery.Selection, base string) int {
	n := 0
	sel.Find("a[href]").Each(func(_ int, a *goquery.Selection) {
		if u, ok := absolute(a, base); ok && sameHost(u, base) {
			n++
		}
	})
	return n
}

// treeFromLists builds a tree from nested lists, or a flat list where the
// navigation has no list markup at all.
func treeFromLists(scope *goquery.Selection, base string, top bool) []*node {
	items := scope.ChildrenFiltered("ul, ol").ChildrenFiltered("li")
	if items.Length() == 0 {
		if !top {
			// A leaf item has no children. Falling through to the flat
			// branch would find the item's own link again and hang a copy of
			// the page beneath itself.
			return nil
		}
		return flatLinks(scope, base)
	}
	var out []*node
	items.Each(func(_ int, li *goquery.Selection) {
		n := itemNode(li, base)
		n.children = treeFromLists(li, base, false)
		out = append(out, n)
	})
	return out
}

// flatLinks is every link in a navigation with no list markup, in document
// order.
func flatLinks(scope *goquery.Selection, base string) []*node {
	var out []*node
	scope.Find("a[href]").Each(func(_ int, a *goquery.Selection) {
		if u, ok := absolute(a, base); ok {
			out = append(out, &node{title: text(a), url: u})
		}
	})
	return out
}

// itemNode is one list item: the page it links to, or the level it names.
func itemNode(li *goquery.Selection, base string) *node {
	if anchor := ownAnchor(li); anchor != nil {
		if u, ok := absolute(anchor, base); ok {
			return &node{title: text(anchor), url: u}
		}
	}
	// A category label with no link of its own still names a level.
	label := li.Find("span, button, div, " + headingSelector).First()
	title := text(li)
	if label.Length() > 0 {
		title = text(label)
	}
	return &node{title: strings.TrimSpace(strings.SplitN(title, "\n", 2)[0])}
}

// ownAnchor is the link belonging to a list item rather than to one of its
// children. Searching descendants alone would let a category item adopt its
// first child's link, taking that child's title and URL and then listing the
// child again beneath itself.
func ownAnchor(li *goquery.Selection) *goquery.Selection {
	var found *goquery.Selection
	li.Find("a[href]").EachWithBreak(func(_ int, a *goquery.Selection) bool {
		holder := a.ParentsFiltered("ul, ol").First()
		if holder.Length() > 0 && !sameNode(holder, li.Parent()) {
			return true
		}
		found = a
		return false
	})
	return found
}

// indexPageTree groups a hub page's links under the headings above them.
//
// A heading inside a link titles that link; a heading outside one opens a
// section. Both shapes appear on the same real page, where cards are written
// as a link wrapping a heading under a section header, and treating every
// heading as a section opener gives each link the title of the one before it.
func indexPageTree(doc *goquery.Document, base string) []*node {
	doc = withoutScripts(doc)
	doc.Find("nav, footer").Remove()
	b := &indexBuilder{base: base, seen: map[string]bool{}}
	root(doc).Find(headingSelector + ", a").Each(func(_ int, el *goquery.Selection) {
		b.visit(el)
	})
	return b.roots
}

// indexBuilder gathers a hub page's sections as its elements arrive.
type indexBuilder struct {
	base  string
	seen  map[string]bool
	roots []*node
	open  []openSection
}

// openSection is a heading still taking links, and the level it opened at.
type openSection struct {
	node  *node
	level int
}

func (b *indexBuilder) visit(el *goquery.Selection) {
	if goquery.NodeName(el) == "a" {
		b.addLink(el)
		return
	}
	if el.ParentsFiltered("a").Length() > 0 {
		// Titles a link; already read with it.
		return
	}
	b.openSection(el)
}

func (b *indexBuilder) addLink(a *goquery.Selection) {
	u, ok := absolute(a, b.base)
	if !ok || !sameHost(u, b.base) || b.seen[u] {
		return
	}
	b.seen[u] = true
	title := text(a)
	if inner := a.Find(headingSelector).First(); inner.Length() > 0 {
		title = text(inner)
	}
	b.add(&node{title: title, url: u})
}

func (b *indexBuilder) openSection(heading *goquery.Selection) {
	level, err := strconv.Atoi(goquery.NodeName(heading)[1:])
	if err != nil {
		return
	}
	for len(b.open) > 0 && b.open[len(b.open)-1].level >= level {
		b.open = b.open[:len(b.open)-1]
	}
	n := &node{title: text(heading)}
	b.add(n)
	b.open = append(b.open, openSection{node: n, level: level})
}

// add hangs a node under the open section, or at the root.
func (b *indexBuilder) add(n *node) {
	if len(b.open) == 0 {
		b.roots = append(b.roots, n)
		return
	}
	parent := b.open[len(b.open)-1].node
	parent.children = append(parent.children, n)
}

// urlPathTree nests pages by their path segments below the seed.
//
// Inferred, with nothing to check it against, which is why Derive warns. It
// still beats a flat list: it keeps a command reference together instead of
// scattering it.
func urlPathTree(urls []string, seed string) []*node {
	baseDepth := len(segmentsOf(seed))
	var roots []*node
	groups := map[string]*node{}
	for _, raw := range urls {
		segments := segmentsOf(raw)
		if len(segments) <= baseDepth {
			roots = append(roots, &node{title: titleFromSlug(pathOf(raw)), url: raw})
			continue
		}
		segments = segments[baseDepth:]
		children := &roots
		prefix := ""
		for _, seg := range segments[:len(segments)-1] {
			prefix += "/" + seg
			group, ok := groups[prefix]
			if !ok {
				group = &node{title: titleFromSlug(seg)}
				groups[prefix] = group
				*children = append(*children, group)
			}
			children = &group.children
		}
		*children = append(*children, &node{
			title: titleFromSlug(segments[len(segments)-1]), url: raw,
		})
	}
	return roots
}

// flatten numbers a tree depth first and keeps the first placement of each
// page: a page with two section numbers is two documents to anything that
// filters by section.
func flatten(nodes []*node, prefix string) []Placement {
	f := &flattener{}
	f.walk(nodes, prefix, nil)
	return f.placements
}

// flattener numbers a tree depth first, keeping the first placement of each
// page: a page with two section numbers is two documents to anything that
// filters by section.
type flattener struct {
	seen       map[string]bool
	placements []Placement
}

func (f *flattener) walk(items []*node, path string, ancestry []string) {
	for i, n := range items {
		section := strconv.Itoa(i + 1)
		if path != "" {
			section = path + "." + section
		}
		f.add(n, section, ancestry)
		next := ancestry
		if n.title != "" {
			next = append(append([]string{}, ancestry...), n.title)
		}
		f.walk(n.children, section, next)
	}
}

func (f *flattener) add(n *node, section string, ancestry []string) {
	if n.url == "" {
		return
	}
	if f.seen == nil {
		f.seen = map[string]bool{}
	}
	if f.seen[n.url] {
		return
	}
	f.seen[n.url] = true
	f.placements = append(f.placements, Placement{
		URL: n.url, Section: section, Ancestry: ancestry, Title: titleOr(n, n.url),
	})
}

// covered is the share of the page set a source places.
func covered(placements []Placement, coverage []string) float64 {
	if len(coverage) == 0 {
		return 0
	}
	placed := make(map[string]bool, len(placements))
	for _, p := range placements {
		placed[p.URL] = true
	}
	n := 0
	for _, u := range coverage {
		if placed[u] {
			n++
		}
	}
	return float64(n) / float64(len(coverage))
}

// titleOr is a node's title, or one read from its URL.
func titleOr(n *node, raw string) string {
	if n.title != "" {
		return n.title
	}
	return titleFromSlug(pathOf(raw))
}

// titleFromSlug turns a path segment into a title.
func titleFromSlug(slug string) string {
	text := strings.Trim(slug, "/")
	if i := strings.LastIndex(text, "/"); i >= 0 {
		text = text[i+1:]
	}
	for _, suffix := range []string{".html", ".htm", ".md"} {
		text = strings.TrimSuffix(text, suffix)
	}
	text = strings.TrimSpace(strings.NewReplacer("-", " ", "_", " ").Replace(text))
	if text == "" {
		return "Untitled"
	}
	return titleCase(text)
}

// titleCase upper-cases each word's first letter, as Python's str.title()
// does for these slugs.
func titleCase(text string) string {
	words := strings.Split(text, " ")
	for i, w := range words {
		if w == "" {
			continue
		}
		words[i] = strings.ToUpper(w[:1]) + strings.ToLower(w[1:])
	}
	return strings.Join(words, " ")
}

// absolute is a link's target, resolved against the page it sits on, when it
// addresses a page at all.
func absolute(a *goquery.Selection, base string) (string, bool) {
	href, ok := a.Attr("href")
	if !ok || !isPageLink(strings.TrimSpace(href)) {
		return "", false
	}
	href = strings.TrimSpace(href)
	baseURL, err := url.Parse(base)
	if err != nil {
		return "", false
	}
	next, err := baseURL.Parse(href)
	if err != nil {
		return "", false
	}
	u, err := fetch.Normalize(next.String())
	if err != nil {
		return "", false
	}
	return u, true
}

// isPageLink reports whether an href addresses a page rather than a
// fragment of this one or another scheme entirely.
func isPageLink(href string) bool {
	if href == "" || strings.HasPrefix(href, "#") {
		return false
	}
	for _, scheme := range []string{"mailto:", "javascript:", "tel:"} {
		if strings.HasPrefix(href, scheme) {
			return false
		}
	}
	return true
}

func sameHost(a, b string) bool {
	ua, err := url.Parse(a)
	if err != nil {
		return false
	}
	ub, err := url.Parse(b)
	if err != nil {
		return false
	}
	return strings.EqualFold(ua.Hostname(), ub.Hostname())
}

func segmentsOf(raw string) []string {
	var out []string
	for _, seg := range strings.Split(pathOf(raw), "/") {
		if seg != "" {
			out = append(out, seg)
		}
	}
	return out
}

func pathOf(raw string) string {
	u, err := url.Parse(raw)
	if err != nil {
		return ""
	}
	return u.Path
}

// text is a node's label, read as the adapters read prose.
func text(sel *goquery.Selection) string {
	if len(sel.Nodes) == 0 {
		return ""
	}
	return html.ProseText(sel.Nodes[0])
}

// withoutScripts drops the elements whose text is not content.
func withoutScripts(doc *goquery.Document) *goquery.Document {
	doc.Find("script, style").Remove()
	return doc
}

// root is the document's body, or the document itself where it has none.
func root(doc *goquery.Document) *goquery.Selection {
	if body := doc.Find("body").First(); body.Length() > 0 {
		return body
	}
	return doc.Selection
}

// sameNode reports whether two selections hold the same element.
func sameNode(a, b *goquery.Selection) bool {
	if len(a.Nodes) == 0 || len(b.Nodes) == 0 {
		return false
	}
	return a.Nodes[0] == b.Nodes[0]
}

// percent renders a share as Python's "{:.0%}" does.
func percent(share float64) string {
	return fmt.Sprintf("%.0f%%", share*100)
}
